### ISSUE TITLE
Testing on Coco and Loss

### ISSUE THREAD

**Issue by anatolix** [michalfaber/keras_Realtime_Multi-Person_Pose_Estimation#12]
Hi. 

I trained model bit further it is still training
<img width="528" alt="screen shot 2017-10-23 at 13 59 58" src="https://user-images.githubusercontent.com/1033375/31886023-4b492860-b7fb-11e7-95c9-86e21ef8ffc1.png">

Any idea what is the target loss should be? (I haven't found in the paper, only COCO results)

**Comment by anatolix** [comment]
Hi.

I tried to get answer for this myself, took converted model **model/keras/model.h5**, renamed it to **weights.best.h5** and loaded it in next stage of training. Numbers show results of this model actually **worse** than my trained model. i.e. 11 vs 1.75 i.e. I mean keras process of training or data preparation is **broken**. Since testing process is obviously works I think it possible the data preparation.

loss: 11.8681 - weight_stage1_L1_loss: 0.0017 - weight_stage1_L2_loss: 0.0012 - weight_stage2_L1_loss: 0.0015 - weight_stage2_L2_loss: 0.0011 - weight_stage3_L1_loss: 0.0014 - weight_stage3_L2_loss: 0.0010 - weight_stage4_L1_loss: 0.0014 - weight_stage4_L2_loss: 0.0010 - weight_stage5_L1_loss: 0.0014 - weight_stage5_L2_loss: 0.0010 - weight_stage6_L1_loss: 0.0014 - weight_stage6_L2_loss: 0.0010 - weight_stage1_L1_acc: 0.0422 - weight_stage1_L2_acc: 0.9749 - weight_stage2_L1_acc: 0.0495 - weight_stage2_L2_acc: 0.9761 - weight_stage3_L1_acc: 0.0479 - weight_stage3_L2_acc: 0.9771 - weight_stage4_L1_acc: 0.8540 - weight_stage4_L2_acc: 0.9776 - weight_stage5_L1_acc: 0.8352 - weight_stage5_L2_acc: 0.9771 - weight_stage6_L1_acc: 0.0500 - weight_stage6_L2_acc: 0.9774 - val_loss: 11.8669 - val_weight_stage1_L1_loss: 0.0015 - val_weight_stage1_L2_loss: 0.0012 - val_weight_stage2_L1_loss: 0.0013 - val_weight_stage2_L2_loss: 0.0011 - val_weight_stage3_L1_loss: 0.0013 - val_weight_stage3_L2_loss: 0.0010 - val_weight_stage4_L1_loss: 0.0013 - val_weight_stage4_L2_loss: 0.0010 - val_weight_stage5_L1_loss: 0.0013 - val_weight_stage5_L2_loss: 0.0010 - val_weight_stage6_L1_loss: 0.0013 - val_weight_stage6_L2_loss: 0.0010 - val_weight_stage1_L1_acc: 0.0454 - val_weight_stage1_L2_acc: 0.9742 - val_weight_stage2_L1_acc: 0.0501 - val_weight_stage2_L2_acc: 0.9752 - val_weight_stage3_L1_acc: 0.0504 - val_weight_stage3_L2_acc: 0.9756 - val_weight_stage4_L1_acc: 0.8673 - val_weight_stage4_L2_acc: 0.9758 - val_weight_stage5_L1_acc: 0.8612 - val_weight_stage5_L2_acc: 0.9759 - val_weight_stage6_L1_acc: 0.0516 - val_weight_stage6_L2_acc: 0.9759

**Comment by michalfaber** [comment]
@anatolix Thanks for trying this path. Indeed, there are some issues with training process in keras.  I will update the code once I find the problem.

**Comment by michalfaber** [comment]
@anatolix Problem solved.

**Comment by anatolix** [comment]
It definitely works very good now. Thank you!

Could you please explain a bit about 'The network never sees the same image twice which was a problem in previous approach (tool rmpe_dataset_transformer)'?

**Comment by michalfaber** [comment]
Exactly. The server performs a random transformation on every image.

**Comment by ul1994** [comment]
I have been running a similar experiment and I am not sure if what I'm seeing is intended. When i train an epoch on top of the downloadable keras trained model `model.h5`, this is the printout:

```bash
loss: 669.5428 - 

weight_stage1_L1_loss: 85.6600 - weight_stage1_L2_loss: 28.8309 - 
weight_stage2_L1_loss: 83.0361 - weight_stage2_L2_loss: 34.7485 - 
weight_stage3_L1_loss: 81.8874 - weight_stage3_L2_loss: 28.2328 - 
weight_stage4_L1_loss: 78.9033 - weight_stage4_L2_loss: 27.0499 - 
weight_stage5_L1_loss: 78.8889 - weight_stage5_L2_loss: 26.4100 - 
weight_stage6_L1_loss: 77.8298 - weight_stage6_L2_loss: 26.2129 - 

weight_stage1_L1_acc: 0.0615 - weight_stage1_L2_acc: 0.9690 - 
weight_stage2_L1_acc: 0.0684 - weight_stage2_L2_acc: 0.9697 - 
weight_stage3_L1_acc: 0.0681 - weight_stage3_L2_acc: 0.9701 - 
weight_stage4_L1_acc: 0.0682 - weight_stage4_L2_acc: 0.9701 - 
weight_stage5_L1_acc: 0.0681 - weight_stage5_L2_acc: 0.9703 - 
weight_stage6_L1_acc: 0.0685 - weight_stage6_L2_acc: 0.9703 - 
```

The losses line up with the screenshot in the readme, but how should one interpret the L1 accuracies? Why are they so low compared to the L2 accuracies?

This was further odd considering with training the model ground up, I saw 50%+ L1 accuracies within the first epoch, leading me to believe something is inconsistent.

```bash
weight_stage1_L1_acc: 0.5193 - weight_stage1_L2_acc: 0.4426 - 
weight_stage2_L1_acc: 0.5447 - weight_stage2_L2_acc: 0.4446 - 
weight_stage3_L1_acc: 0.5456 - weight_stage3_L2_acc: 0.4404 - 
weight_stage4_L1_acc: 0.5447 - weight_stage4_L2_acc: 0.4436 - 
weight_stage5_L1_acc: 0.5490 - weight_stage5_L2_acc: 0.4427 - 
weight_stage6_L1_acc: 0.5466 - weight_stage6_L2_acc: 0.4427 
```

**Comment by piperod** [comment]
I am getting a nan in the loss function. Any idea on what I am doing wrong? Accuracy is reporting fine.

 loss: nan - weight_stage1_L1_loss: nan - weight_stage1_L2_loss: nan - weight_stage2_L1_loss: nan - weight_stage2_L2_loss: nan - weight_stage3_L1_loss: nan - weight_stage3_L2_loss: nan - weight_stage4_L1_loss: nan - weight_stage4_L2_loss: nan - weight_stage5_L1_loss: nan - weight_stage5_L2_loss: nan - weight_stage6_L1_loss: nan - weight_stage6_L2_loss: nan -

**Comment by ksaluja15** [comment]
@piperod assuming everything else is correct, did u increase the base learning rate ? Increasing the base lr leads to loss being nan

**Comment by piperod** [comment]
@ksaluja15 thanks for replying.  I didn't touch the learning rate. However  I did modify the parts since I don't have 17 annotations. Is this code  in line 94 at generate_hdf5.py, related with left [part,0], right [part,1] and  [partv,0] visibility?  I think also my problem is that I might have to modify  the augmenter too. 

````
pers["joint"] = np.zeros((17, 3))
                for part in range(17):
                    pers["joint"][part, 0] = anno[part * 3]
                    pers["joint"][part, 1] = anno[part * 3 + 1]
                    if anno[part * 3 + 2] == 2:
                        pers["joint"][part, 2] = 1
                    elif anno[part * 3 + 2] == 1:
                        pers["joint"][part, 2] = 0
                    else:
                        pers["joint"][part, 2] = 2
````

**Comment by anatolix** [comment]
@piperod I've made fork with my custom augmenter https://github.com/anatolix/keras_Realtime_Multi-Person_Pose_Estimation and stuck into NaN problem too. Trying to solve now.

The idea of augmentation is the following, if you just delete some parts you could delete proper heatmaps and pafs. If you add or change parts you need new heatmaps and pafs

**Comment by piperod** [comment]
@anatolix thanks a lot,  I will try your code asap. I was able to get ride of the nans, but I thinks is a very naive solution. I just basically put several copies of my 5 annotations in the 17 spots available in generate_hdf5.py lines 96-100. Still don't have results, but I guess maybe having  a code more flexible  in the number of annotations could be a good feature to add in the future.

**Comment by anatolix** [comment]
About accuracy: keras use binary classification accuracy if you just ask for 'accuracy' this metric is not relevant, I think it convert floats in pafs and heatmaps to booleans and compares them.  It is nothing like COCO or MPII accuracies which mentioned in original paper.

**Comment by piperod** [comment]
It is weird, it went well for about 500 epochs, then nan is reported again.

**Comment by piperod** [comment]
@anatolix You were right. The fact that I was hardcoding wrongly the new parts were causing to get these divisions by 0. However I was able to modify the joints and the mapping for the skeletons (datatransformer.cpp)  to match my annotations and so far is working. The loss has dropped from 2000 to 145 in the first 1600 epochs. Haven't test yet the model, but seems like a good start. I still figuring out what was the error with the python code you provided, as soon as I have results will let you know. Thanks for the patience and the help.

**Comment by piperod** [comment]
Just a little update. Its working great! Loss function is working now, and predictions are close to what is shown in the demo. 
![captura de pantalla 2017-11-14 a la s 11 30 15 p m](https://user-images.githubusercontent.com/13397888/32819045-07eb3a12-c995-11e7-95fe-8034d1cf02a0.png)
